shared/task.rs

```diff
@@ -1,89 +1,404 @@
+        }
+    }
+
+    pub fn is_paused(&self) -> bool {
+        match self.state {
+            TaskState::Running {
+                status: Running::Paused,
+                ..
+            } => true,
+            _ => false,
+        }
+    }
+
+
+    pub fn is_delayed(&self) -> bool {
+        match self.state {
+            TaskState::Waiting {
+                enqueue_at: Some(_),
+                ..
+            } => true,
+            _ => false,
+        }
     }
 
     pub fn is_queued(&self) -> bool {
-        return self.status == TaskStatus::Queued || self.status == TaskStatus::Stashed;
+        match &self.state {
+            TaskState::Waiting { status: Waiting::Queued, .. } => true,
+            _ => false,
+        }
+    }
+}
+
+// Get general info about task in it's context
+pub enum Startability {
+    Ready,
+    Waiting(Duration),
+    Dependencies(Vec<usize>),
+    DependenciesFailure(usize),
+    Stashed,
+    Locked,
+
+    Unknown,
+}
+
+pub enum GeneralState {
+    Healthy,
+    Failed,
+    Paused,
+    Waiting,
+}
+
+impl Task {
+    pub fn can_be_started(&self, state: &State) -> bool {
+        match self.start_info(state) {
+            Startability::Ready => true,
+            _ => false,
+        }
+    }
+
+    pub fn start_info(&self, state: &State) -> Startability {
+        match &self.state {
+            TaskState::Waiting {
+                locked: true, ..
+            } => Startability::Locked,
+
+            TaskState::Waiting {
+                status: Waiting::Stashed,
+                ..
+            } => Startability::Stashed,
+
+            TaskState::Waiting {
+                status: Waiting::Queued,
+                enqueue_at: Some(time),
+                ..
+            } if time > &Local::now() => {
+                Startability::Waiting(*time - Local::now())
+            }
+
+            TaskState::Waiting {
+                status: Waiting::Queued,
+                dependencies,
+                ..
+            } => {
+                let dep_not_done: Vec<_> = dependencies
+                    .iter()
+                    .flat_map(|id| state.get_task(*id))
+                    .filter(|task| !task.is_done())
+                    .collect();
+
+                if !dep_not_done.is_empty() {
+                    let failed = dep_not_done.iter().find(|task| task.is_errored());
+
+                    if let Some(failed_task) = failed {
+                        Startability::DependenciesFailure(failed_task.id)
+                    } else {
+                        Startability::Dependencies(dep_not_done.iter().map(|task| task.id).collect())
+                    }
+                } else {
+                    Startability::Ready
+                }
+            }
+            _ => Startability::Unknown,
+        }
+    }
+
+    pub fn general_state(&self, state: &State) -> GeneralState {
+        match &self.state {
+            TaskState::Waiting { .. } => match self.start_info(state) {
```

Comment:
Without DependenciesFailure and Unknown (which IMO is really bad practice) the whole start_info stuff can be removed.
The only place this should be used in, is in the TaskHandler, since that's the only place where we have to check whether a task can be started or not.